File: server-go/keyboard_exemption_test.go
package main

import "testing"

// The keyboard-only accessibility exemption and the key-hold check that keeps a
// keyboard-driven agent from claiming it. See keyboardOnlyUser in scoring.go.

func keyboardSignals(keyEvents float64, extra map[string]interface{}) map[string]interface{} {
	b := map[string]interface{}{"totalPoints": 0.0, "trajectoryLength": 0.0, "keyEvents": keyEvents, "touchEvents": 0.0}
	for k, v := range extra {
		b[k] = v
	}
	return map[string]interface{}{
		"behavioral":    b,
		"environmental": map[string]interface{}{"automationFlags": map[string]interface{}{}},
	}
}

func TestKeyboardExemptionStandsWithoutHoldData(t *testing.T) {
	// An older widget, or a visitor who tabbed to the checkbox and pressed
	// Space without typing into a field, reports counts and nothing else.
	sig := keyboardSignals(8, nil)
	if !keyboardOnlyUser(sig, getMap(sig, "behavioral")) {
		t.Error("counts alone must keep the exemption")
	}
}

func TestKeyboardExemptionStandsForFingers(t *testing.T) {
	sig := keyboardSignals(12, map[string]interface{}{"keyHoldSamples": 6.0, "keyHoldAvg": 85.0})
	if !keyboardOnlyUser(sig, getMap(sig, "behavioral")) {
		t.Error("85ms holds are a hand on a key")
	}
}

func TestKeyboardExemptionDeniedForMechanicalHolds(t *testing.T) {
	sig := keyboardSignals(12, map[string]interface{}{"keyHoldSamples": 6.0, "keyHoldAvg": 4.0})
	if keyboardOnlyUser(sig, getMap(sig, "behavioral")) {
		t.Error("4ms holds are an automation protocol, not a keyboard user")
	}
}

func TestKeyboardExemptionPoolsFormFieldDwell(t *testing.T) {
	// The form analyser's per-field dwell times count too, so a widget that
	// typed into a field is judged even without the session-level summary.
	mechanical := keyboardSignals(9, nil)
	mechanical["formAnalysis"] = map[string]interface{}{"textareaKeyboard": map[string]interface{}{
		"message": map[string]interface{}{"dwellTimes": []interface{}{2.0, 3.0, 1.0, 2.0}},
	}}
	if keyboardOnlyUser(mechanical, getMap(mechanical, "behavioral")) {
		t.Error("mechanical field dwell must deny the exemption")
	}
	human := keyboardSignals(9, nil)
	human["formAnalysis"] = map[string]interface{}{"textareaKeyboard": map[string]interface{}{
		"message": map[string]interface{}{"dwellTimes": []interface{}{60.0, 75.0, 90.0}},
	}}
	if !keyboardOnlyUser(human, getMap(human, "behavioral")) {
		t.Error("human field dwell must keep the exemption")
	}
}

func TestKeyboardExemptionNeedsEnoughHoldsToJudge(t *testing.T) {
	sig := keyboardSignals(4, map[string]interface{}{"keyHoldSamples": 2.0, "keyHoldAvg": 3.0})
	if !keyboardOnlyUser(sig, getMap(sig, "behavioral")) {
		t.Error("two holds are too few to take an exemption away on")
	}
}

func TestKeyboardExemptionRequiresKeysAndNoPointer(t *testing.T) {
	one := keyboardSignals(1, nil)
	if keyboardOnlyUser(one, getMap(one, "behavioral")) {
		t.Error("one key event is not keyboard use")
	}
	moved := keyboardSignals(8, map[string]interface{}{"totalPoints": 3.0})
	if keyboardOnlyUser(moved, getMap(moved, "behavioral")) {
		t.Error("a visitor who moved the pointer is not keyboard-only")
	}
}

// End to end: a keyboard-driven agent is scored as pointerless — both movement
// views fire and corroborate — while a keyboard-only person is not.
func TestKeyboardAgentIsScoredAsPointerless(t *testing.T) {
	e := NewScoringEngine("test-secret")
	agent := keyboardSignals(14, map[string]interface{}{"keyHoldSamples": 14.0, "keyHoldAvg": 2.0})
	dets := append(e.detectVisionAI(agent), e.detectBehavioral(agent)...)
	if !hasReasonContaining(dets, "Zero mouse, touch, or keyboard events") {
		t.Errorf("agent should lose the exemption and score as pointerless, got %+v", dets)
	}
	if !hasReasonContaining(dets, "No mouse movement detected before click") {
		t.Errorf("agent should trip the vision_ai movement check, got %+v", dets)
	}
	if got := applyCorroborationFloor(0.1, dets); got < corroborationFloor {
		t.Errorf("two movement views should corroborate to %v, got %v", corroborationFloor, got)
	}

	person := keyboardSignals(14, map[string]interface{}{"keyHoldSamples": 14.0, "keyHoldAvg": 80.0})
	dets = append(e.detectVisionAI(person), e.detectBehavioral(person)...)
	if hasReasonContaining(dets, "Zero mouse") || hasReasonContaining(dets, "No mouse movement") {
		t.Errorf("a keyboard-only person must keep the exemption, got %+v", dets)
	}
}

File: server-node/keyboard-exemption.test.js
'use strict';

// The keyboard-only accessibility exemption and the key-hold check that keeps a
// keyboard-driven agent from claiming it. Mirrors server-go/keyboard_exemption_test.go.

const { test } = require('node:test');
const assert = require('node:assert/strict');
const {
  keyboardOnlyUser, detectVisionAI, detectBehavioral, applyCorroborationFloor, CORROBORATION_FLOOR,
} = require('./engine');

const kb = (keyEvents, extra = {}) => ({
  behavioral: { totalPoints: 0, trajectoryLength: 0, keyEvents, touchEvents: 0, ...extra },
  environmental: { automationFlags: {} },
});
const reasons = (dets) => dets.map((d) => d.reason).join(' | ');

test('the exemption stands without hold data', () => {
  const s = kb(8);
  assert.equal(keyboardOnlyUser(s, s.behavioral), true);
});

test('the exemption stands for fingers', () => {
  const s = kb(12, { keyHoldSamples: 6, keyHoldAvg: 85 });
  assert.equal(keyboardOnlyUser(s, s.behavioral), true);
});

test('the exemption is denied for mechanical holds', () => {
  const s = kb(12, { keyHoldSamples: 6, keyHoldAvg: 4 });
  assert.equal(keyboardOnlyUser(s, s.behavioral), false);
});

test('form-field dwell times are pooled into the judgement', () => {
  const mechanical = { ...kb(9), formAnalysis: { textareaKeyboard: { message: { dwellTimes: [2, 3, 1, 2] } } } };
  assert.equal(keyboardOnlyUser(mechanical, mechanical.behavioral), false);
  const human = { ...kb(9), formAnalysis: { textareaKeyboard: { message: { dwellTimes: [60, 75, 90] } } } };
  assert.equal(keyboardOnlyUser(human, human.behavioral), true);
});

test('too few holds cannot take the exemption away', () => {
  const s = kb(4, { keyHoldSamples: 2, keyHoldAvg: 3 });
  assert.equal(keyboardOnlyUser(s, s.behavioral), true);
});

test('the exemption requires key events and no pointer', () => {
  const one = kb(1);
  assert.equal(keyboardOnlyUser(one, one.behavioral), false);
  const moved = kb(8, { totalPoints: 3 });
  assert.equal(keyboardOnlyUser(moved, moved.behavioral), false);
});

test('a keyboard-driven agent is scored as pointerless and corroborates; a keyboard person is not', () => {
  const agent = kb(14, { keyHoldSamples: 14, keyHoldAvg: 2 });
  let dets = [...detectVisionAI(agent), ...detectBehavioral(agent)];
  assert.match(reasons(dets), /Zero mouse, touch, or keyboard events/);
  assert.match(reasons(dets), /No mouse movement detected before click/);
  assert.ok(applyCorroborationFloor(0.1, dets) >= CORROBORATION_FLOOR);

  const person = kb(14, { keyHoldSamples: 14, keyHoldAvg: 80 });
  dets = [...detectVisionAI(person), ...detectBehavioral(person)];
  assert.doesNotMatch(reasons(dets), /Zero mouse|No mouse movement/);
});
